src/java/org/apache/cassandra/index/accord/RangeMemoryIndex.java

```diff
@@ -58,13 +63,51 @@ public class RangeMemoryIndex
 {
 
     @GuardedBy("this")
-    private final Map<Group, RangeTree<byte[], Range, DecoratedKey>> map = new HashMap<>();
-    @GuardedBy("this")
-    private final Map<Group, Metadata> groupMetadata = new HashMap<>();
+    private final Map<Key, Group> map = new HashMap<>();
 
-    private static class Metadata
+    private static class Group
     {
+        private RangeTree<byte[], Range, DecoratedKey> tree = createRangeTree();
         public byte[] minTerm, maxTerm;
+        public TxnId minTimestamp = TxnId.MAX;
+        public TxnId maxTimestamp = TxnId.NONE;
+
+        void add(Range range, DecoratedKey key, TxnId txnId, byte[] start, byte[] end)
+        {
+            tree.add(range, key);
+            minTerm = minTerm == null ? start : ByteArrayUtil.compareUnsigned(minTerm, 0, start, 0, minTerm.length) > 0 ? start : minTerm;
+            maxTerm = maxTerm == null ? end : ByteArrayUtil.compareUnsigned(maxTerm, 0, end, 0, maxTerm.length) < 0 ? end : maxTerm;
+            if (minTimestamp.compareTo(txnId) > 0)
+                minTimestamp = txnId;
+            if (maxTimestamp.compareTo(txnId) < 0)
+                maxTimestamp = txnId;
+        }
+
+        void search(byte[] start, byte[] end,
+                    Timestamp minTimestamp, Timestamp maxTimestamp,
+                    Consumer<Map.Entry<RangeMemoryIndex.Range, DecoratedKey>> fn)
+        {
+            if (this.minTimestamp.compareTo(maxTimestamp) > 0 || this.maxTimestamp.compareTo(minTimestamp) < 0)
+                return;
+            tree.search(new Range(start, end), e -> {
+                TxnId id = AccordKeyspace.JournalColumns.getJournalKey(e.getValue()).id;
```

Comment:
not a fan of this cost, but my thinking is as follows

1) the in-memory cost is the `ByteBuffer`
2) only called when the range intersects, so each call are limited to possible matches, so just need to do the timestamp filter.

I don't know if min/max are no-op in the common case (i doubt it as it should be based off durability as far as i can tell), so don't think it really makes sense to try to optimize the filter to avoid this cost when min/max are the full range

Reply:
2 things i can do to lower this cost

1) rather than construct the full `JournalColum`, only construct the `TxnId`.  This "should" be very similar CPU cost but would have slightly less memory cost
2) try to do a byte compare to avoid the memory cost.  I don't believe we have this setup already so think i would have to roll my own here.

For this class i don't think it matters too much (its the stuff that hasn't been flushed to sstable), but for the sstable side this "should" matter more i think

Reply:
Ok so think i can just remove this logic...

`org.apache.cassandra.service.accord.RangeSearcher.DefaultResult#consume(java.util.function.Consumer<accord.primitives.TxnId>)`

We filter at this layer so won't return the result to the caller without applying the filter... but this means we have to deserialize the key to do the filter, so this logic just does that twice!

Reply:
FYI given how we do things today, we don't ever touch this object... journal directly creates SSTables and ignores memtable, so it really doesn't matter too much what we do here...

2i *requires* we have memory indexes and its mostly here just to be "correct" if we ever do rely on memtable for some reason (such as to patch a journal entry?)...